kernel/init.c

```diff
@@ -311,24 +311,19 @@ static int do_device_init(const struct device *dev)
 {
 	int rc = 0;
 
+	/* device initialization has been invoked */
+	dev->state->initialized = true;
+
 	if (dev->ops.init != NULL) {
 		rc = dev->ops.init(dev);
-		/* Mark device initialized. If initialization
-		 * failed, record the error condition.
-		 */
+		/* If initialization failed, record the error condition. */
 		if (rc != 0) {
-			if (rc < 0) {
-				rc = -rc;
-			}
-			if (rc > UINT8_MAX) {
-				rc = UINT8_MAX;
-			}
-			dev->state->init_res = rc;
+			int res = (rc < 0) ? -rc : rc;
+
+			dev->state->init_res = (res <= UINT8_MAX) ? res : UINT8_MAX;
```

Comment:
Two things jump out at me here.

The first is that unless I am missing something, this just gives the same behavior as the original code. If that is true, how does that fix anything?

The second is a logical corner case. If the return code `rc` from `dev->ops.init()` is ever INT_MIN (0x80000000), then `-rc` is the same as `rc`, and the value that we save will be 0x0--a false success path. Would it be worthwhile to have an assert that covers that?

Reply:
Aaah ... I finally see what my eyes kept glazing over on the previous pass with regards to the signedness of `rc`.